src/python/pants/backend/codegen/protobuf/python/rules.py

```diff
@@ -47,13 +64,53 @@ async def generate_python_from_protobuf(
     # Protoc needs all transitive dependencies on `protobuf_libraries` to work properly. It won't
     # actually generate those dependencies; it only needs to look at their .proto files to work
     # with imports.
-    transitive_targets = await Get(TransitiveTargets, Addresses([request.protocol_target.address]))
+    # TODO(#10917): This monstrosity is because we are not able to use
+    # `await Get(TransitiveTargets`) without causing rule graph issues. So, we copy a hacky
+    # implementation of the rules to resolve direct deps + transitive deps. This implementation is
+    # much less robust:
+    #
+    #    - Does not use dependency injection.
+    #    - Does not use dependency inference.
+    #    - Always places dependencies on subtargets, and isn't as careful about avoid self-cycles.
+    #    - Worse performance, as this doesn't batch as efficiently.
+    #
+    # Normally, these restrictions would be a non-starter, but because we are solely looking for
+    # transitive dependencies on Protobuf libraries, these hacks are tolerable for now.
+    visited: OrderedSet[Target] = OrderedSet()
+    queued = OrderedSet([request.protocol_target])
+    while queued:
+        tgt = queued.pop()
+        visited.add(tgt)
```

Comment:
A new user is looking to implement Cython support via codegen, so #10917 is going to hit them. In the meantime to us fixing it, I'm going to factor up these Dependency Lite and TransitiveTargets Lite implementations so that the user can use them in their plugin. They will only need to `s/TransitiveTargets/TransitiveTargetsLite`.

From my testing in #10917, the key difference with these Lite implementations is that they will not implement dep inference support. I can add back injection support, and can fix the performance regression. This should be temporarily fine because neither Cython nor Protobuf have dep inference yet.

(To be clear, this is all a short-term holdover. We need to fix the graph.)

Reply:
SG